crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,410 @@
+    # a hash?
+    def remove_pending_elements(bc, inst, elements)
+      nodes_map = elements_to_nodes_to_roles_map(elements)
+
+      # Remove the entries from the nodes.
+      f = file_lock.acquire("BA-LOCK", logger: logger)
+      begin
+        nodes_map.each do |node_name, data|
+          node = NodeObject.find_node_by_name(node_name)
+          next if node.nil?
+          unless node.crowbar["crowbar"]["pending"].nil? or node.crowbar["crowbar"]["pending"]["#{bc}-#{inst}"].nil?
+            node.crowbar["crowbar"]["pending"]["#{bc}-#{inst}"] = {}
+            node.save
+          end
+        end
+      ensure
+        file_lock.release(f)
+      end
+    end
+
+    # Create map with nodes and their element list
+    # Transform ( {role => [nodes], role1 => [nodes]} hash to { node => [roles], node1 => [roles]},
+    # accounting for clusters
+    def elements_to_nodes_to_roles_map(elements)
+      nodes_map = {}
+
+      elements.each do |role_name, nodes|
+        # Expand clusters to individual nodes
+        nodes, failures = expand_nodes_for_all(nodes)
+        unless failures.nil? || failures.empty?
+          logger.debug "elements_to_nodes_to_roles_map: skipping items that we failed to expand: #{failures.join(", ")}"
+        end
+
+        # Add the role to node's list
+        nodes.each do |node_name|
+          if NodeObject.find_node_by_name(node_name).nil?
+            logger.debug "elements_to_nodes_to_roles_map: skipping deleted node #{node_name}"
+            next
+          end
+          nodes_map[node_name] = [] if nodes_map[node_name].nil?
+          nodes_map[node_name] << role_name
+        end
+      end
+
+      nodes_map
+    end
+
+    # Get a hash of {node => [roles], node1 => [roles]}
+    def add_pending_elements(bc, inst, elements, queue_me, pre_cached_nodes = {})
+      nodes_map = elements_to_nodes_to_roles_map(elements)
+
+      # We need to be sure that we're the only ones modifying the node records at this point.
+      # This will work for preventing changes from rails app, but not necessarily chef.
+      # Tough luck.
+      f = file_lock.acquire("BA-LOCK", logger: logger)
+
+      # Delay is the list of nodes that are not ready and are needed for this deploy to run
+      delay = []
+      pre_cached_nodes = {}
+      begin
+        # Check for delays and build up cache
+        # FIXME: why?
+        if queue_me
+          delay = nodes_map.keys
+        else
+          delay, pre_cached_nodes = elements_not_ready(nodes_map.keys, pre_cached_nodes)
+        end
+
+        # We have all nodes we'll ever need. Mark them as applying and this proposal
+        # as the 'author'
+        if delay.empty?
+          nodes_map.each do |node_name, val|
+            node = pre_cached_nodes[node_name]
+
+            # Nothing to wait for so mark them applying now.
+            node.crowbar['state'] = 'applying'
+            node.crowbar['state_owner'] = "#{bc}-#{inst}"
```

Comment:
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.